# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webgpu/Device.cpp
# Commit: 7f25d62e49a2
# Full Hash: 7f25d62e49a2325004433386f09955d3c2f64c3c
# Author: sotaro <sotaro.ikeda.g@gmail.com>
# Date: 2023-10-24 09:38:27
# Regressor Bug: 1856787
# File Overlap Count: 2
# Description:
#   Bug 1856787 - Add a capability to present WebGPU without readback on Windows r=webgpu-reviewers,nical
#   
#   For presenting WebGPU without readback, wgpu does rendering to ExternalTexture. Then the ExternalTexture is pushed to RemoteTextureMap for present.
#   
#   With DX12, ExternalTextureD3D11 is implemented for gecko side implementation. ExternalTextureWgpu holds necessary resource that is necessary by wgpu. ExternalTextureWgpu is created and destroyed by gecko side's ExternalTexture.
# ==============================================================================

diff -r ed56525c65bc -r 7f25d62e49a2 dom/webgpu/Device.cpp
--- a/dom/webgpu/Device.cpp	Tue Oct 24 03:54:42 2023 +0000
+++ b/dom/webgpu/Device.cpp	Tue Oct 24 04:04:07 2023 +0000
@@ -148,6 +148,26 @@
   return Buffer::Create(this, mId, aDesc, aRv);
 }
 
+already_AddRefed<Texture> Device::CreateTextureForSwapChain(
+    const dom::GPUCanvasConfiguration* const aConfig,
+    const gfx::IntSize& aCanvasSize, layers::RemoteTextureOwnerId aOwnerId) {
+  MOZ_ASSERT(aConfig);
+
+  dom::GPUTextureDescriptor desc;
+  desc.mDimension = dom::GPUTextureDimension::_2d;
+  auto& sizeDict = desc.mSize.SetAsGPUExtent3DDict();
+  sizeDict.mWidth = aCanvasSize.width;
+  sizeDict.mHeight = aCanvasSize.height;
+  sizeDict.mDepthOrArrayLayers = 1;
+  desc.mFormat = aConfig->mFormat;
+  desc.mMipLevelCount = 1;
+  desc.mSampleCount = 1;
+  desc.mUsage = aConfig->mUsage | dom::GPUTextureUsage_Binding::COPY_SRC;
+  desc.mViewFormats = aConfig->mViewFormats;
+
+  return CreateTexture(desc, Some(aOwnerId));
+}
+
 already_AddRefed<Texture> Device::CreateTexture(
     const dom::GPUTextureDescriptor& aDesc) {
   return CreateTexture(aDesc, /* aOwnerId */ Nothing());
@@ -328,10 +348,12 @@
 }
 
 already_AddRefed<Texture> Device::InitSwapChain(
-    const dom::GPUCanvasConfiguration& aDesc,
+    const dom::GPUCanvasConfiguration* const aConfig,
     const layers::RemoteTextureOwnerId aOwnerId,
     bool aUseExternalTextureInSwapChain, gfx::SurfaceFormat aFormat,
     gfx::IntSize aCanvasSize) {
+  MOZ_ASSERT(aConfig);
+
   if (!mBridge->CanSend()) {
     return nullptr;
   }
@@ -342,20 +364,9 @@
   mBridge->DeviceCreateSwapChain(mId, rgbDesc, maxBufferCount, aOwnerId,
                                  aUseExternalTextureInSwapChain);
 
-  dom::GPUTextureDescriptor desc;
-  desc.mDimension = dom::GPUTextureDimension::_2d;
-  auto& sizeDict = desc.mSize.SetAsGPUExtent3DDict();
-  sizeDict.mWidth = aCanvasSize.width;
-  sizeDict.mHeight = aCanvasSize.height;
-  sizeDict.mDepthOrArrayLayers = 1;
-  desc.mFormat = aDesc.mFormat;
-  desc.mMipLevelCount = 1;
-  desc.mSampleCount = 1;
-  desc.mUsage = aDesc.mUsage | dom::GPUTextureUsage_Binding::COPY_SRC;
-  desc.mViewFormats = aDesc.mViewFormats;
   // TODO: `mColorSpace`: <https://bugzilla.mozilla.org/show_bug.cgi?id=1846608>
   // TODO: `mAlphaMode`: <https://bugzilla.mozilla.org/show_bug.cgi?id=1846605>
-  return CreateTexture(desc, Some(aOwnerId));
+  return CreateTextureForSwapChain(aConfig, aCanvasSize, aOwnerId);
 }
 
 bool Device::CheckNewWarning(const nsACString& aMessage) {